### PR TITLE
Ensure auto-transformed variables use specified starting values

### LINKED ISSUE #2042: Start value doesn't work for Uniform distribution
It looks like when a variable has a `Uniform` distribution, its `start` value is ignored. This example shows that `alpha` never gets initialized to 5 when it has a `Uniform` distribution, but it does when it has a `Normal` distribution.

```python
import numpy as np
import pymc3 as pm
import theano

# Initialize random number generator
np.random.seed(123)

# True parameter values
alpha, sigma = 1, 1
beta = [1, 2.5]

# Size of dataset
size = 100

# Predictor variable
X1 = np.random.randn(size)
X2 = np.random.randn(size) * 0.2

# Simulate outcome variable
Y = alpha + beta[0]*X1 + beta[1]*X2 + np.random.randn(size)*sigma

# Specify model
@theano.as_op([theano.tensor.dscalar, theano.tensor.dscalar, theano.tensor.dscalar], [theano.tensor.dvector])
def _mu(alpha, beta_0, beta_1):
    print(alpha, beta_0, beta_1)
    return alpha + beta_0*X1 + beta_1*X2

basic_model = pm.Model()

with basic_model:
    alpha = pm.Uniform('alpha', -10, 10)
    #alpha = pm.Normal('alpha', mu=0, sd=10)
    beta_0 = pm.Normal('beta_0', mu=0, sd=10)
    beta_1 = pm.Normal('beta_1', mu=0, sd=10)
    sigma = pm.HalfNormal('sigma', sd=1)
    mu = pm.Deterministic('mu', _mu(alpha, beta_0, beta_1))
    Y_obs = pm.Normal('Y_obs', mu=mu, sd=sigma, observed=Y)

with basic_model:
    step = pm.Metropolis([alpha, beta_0, beta_1, sigma])
    start = {'alpha': 5., 'beta_0': 4., 'beta_1': 3.}
    trace = pm.sample(10, step=step, start=start, init=None)
```

### PR DESCRIPTION
**Pull request by fonnesbeck**
Currently starting values specified for variables that are auto-transformed are ignored by `sample`. This adds a function `_transformed_init` which ensures these values are properly transformed and included in the `start` dict.

Closes #2042

### CODE REVIEW

**Comment by fonnesbeck** [comment]
This turned out to be much harder than I had anticipated. Interval transforms require bound information in order to properly transform values. This update now encodes the bound info into the transformed variable name, if it is a bounded or interval transform. So, for example:

    alpha_interval_

now becomes:

    alpha_interval_(-10,10)_

for example. I'm not sure I like this, but I don't see any easy way to supply this information so that starting values can be properly transformed. Please review.

cc @twiecki

File: pymc3/model.py [hunk 1]
```diff
@@ -994,6 +994,14 @@ def __init__(self, type=None, owner=None, index=None, name=None,
...
             transformed_name = "{}_{}_".format(name, transform.name)
+            bound_dict = {'a':None, 'b':None}
+            for key in bound_dict:
+                try:
+                    bound_dict[key] = transform.__dict__[key].eval()
+                except KeyError:
+                    pass
+            if bound_dict['a'] is not None or bound_dict['b'] is not None:
+                transformed_name += '({},{})_'.format(bound_dict['a'], bound_dict['b'])                  
             self.transformed = model.Var(
```
Comment (twiecki) [inline pymc3/model.py]: Can't we monkey-patch `lower` and `upper` onto `self.transformed`? While not clean, it would be cleaner than this.

**Comment by twiecki** [comment]
I have no problem adding this to the name, but a problem with reading it out. This is definitely a point where we should take a step back and think how we painted ourselves into this corner.

**Comment by fonnesbeck** [comment]
I agree. Probably time to overhaul the initialization procedure.

**Comment by fonnesbeck** [comment]
OK, this should suffice. I have just pushed the `model` from the context into `_soft_update` so that transformation can be extracted. The `transform` instance is now an attribute of `TransformedRV` so that it can be retrieved as needed.

**Comment by fonnesbeck** [comment]
Test failure on G-R statistic, which does not make sense.

**Comment by fonnesbeck** [comment]
Curious that this fails only on 2.7

**Comment by fonnesbeck** [comment]
Fixed now. Please squash-merge if we are happy with it.

File: pymc3/blocking.py [hunk 2]
```diff
@@ -91,6 +94,98 @@ def mapf(self, f):
         return Compose(f, self.rmap)
 
 
+class ListArrayOrdering(object):
```
Comment (twiecki) [inline pymc3/blocking.py]: This shouldn't show up.

**Comment by fonnesbeck** [comment]
Fixed.

**Comment by twiecki** [comment]
Not sure I see how this fixes the issue?

**Comment by fonnesbeck** [comment]
It transforms initial values specified on the original scale and inserts them into the start dict for the transformed variable.

**Comment by fonnesbeck** [comment]
There is an example in the test.

**Comment by twiecki** [comment]
I get it now. Much cleaner than before, the only nit-pick is that `_soft_update` was a general purpose function before and is now tied to the model. Not a big deal though, we can merge.

**Comment by fonnesbeck** [comment]
I can revert it to a general function and decorate it if you like.

**Comment by twiecki** [comment]
Is the function used anywhere else? Maybe we can just rename it.

**Comment by fonnesbeck** [comment]
Not used elsewhere. I've renamed it.

**Comment by twiecki** [comment]
Cool, LGTM.

**Pull request merged**